_posts/2025_01_08_exploiting_capabilities.markdown

```diff
@@ -0,0 +1,73 @@
+                                             uint8_t     qos,
+                                             const char *topic,
+                                             size_t      topicLength,
+                                             const void *payload,
+                                             size_t      payloadLength,
+                                             bool        retain = false);
+```
+
+and has the following comment about their scope:
+
+_Both the topic and payload buffers must remain valid during the execution of this function._
+_If the caller frees them during the execution of this function, the publish may leak application data to the broker through the topic and payload._
+
+Ok, so it seems clear that I only need to make sure that they remain valid and unchanged until the call returns, but I was hitting a bug that seemed to suggest they might be being used outside of that by another thread in the network stack.
+_It turns out I was wrong about that BTW, but anyway I'm a suspicious kind of guy when it comes to trusting external modules._
+A quick look though the code quickly persuaded me that a) I'm not very good at reading other peoples code and b) it was probably not the issue but I still wanted to rule it out.
+
+Luckily by thinking of these in terms of capabilities rather than pointers I can directly implement my intent rather that trust what the library is telling me.
+Specifically I can derive new capabilities that are read only and can't be captured.
+So I can be guaranteed that at the end of the call nothing in the network stack can have modified the message or saved a pointer to in.
+And for good measure I can also limit the bounds so that I don't have to trust topicLength and payloadLength will be used properly.
+This is what the resulting code looks like in my demo:
+
+```c++
+// Publish a string to the status topic
+void publish(SObj mqtt, std::string topic, void *status, bool retain)
+{
+	// Use capabilities with only Load
+	// permission so we can be sure the MQTT stack
+	// doesn't capture a pointer to our buffer
+	Timeout           t{MS_TO_TICKS(5000)};
+	
+	CHERI::Capability roTopic{topic.data()};
+	roTopic.permissions() &= {CHERI::Permission::Load};
+	roTopic.bounds() = topic.size();
+
+	CHERI::Capability roStatus{status};
+	roStatus.permissions() &= {CHERI::Permission::Load};
+
+	auto ret = mqtt_publish(&t,
+	                        mqtt,
+	                        1, // QoS 1 = delivered at least once
+	                        roTopic,
+	                        roTopic.bounds(),
+	                        roStatus,
+	                        roStatus.bounds(),
+							retain);
+
+	if (ret < 0)
+	{
+		Debug::log("Failed to send status: error {}", ret);
+	}
+}
+
+```
+
+So by explicitly using the extra schematics that CHERIoT provides we can remove a trust relationship between our code and some third party module, and to me that's a big step forwards in creating safe and reliable systems. 
```

Comment:
It would be nice to be clear that these guarantees exist *without* trusting anything in their code, including their compiler. If you rely on language-level things then malicious code could exploit a bug in the compiler or could use escape hatches such as `unsafe`.

Reply:
Done